core/state/statedb.go

```diff
@@ -767,6 +786,21 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 	// Finalise all the dirty storage states and write them into the tries
 	s.Finalise(deleteEmptyObjects)
 
+	// Now we're about to start to write changes to the trie. The trie is so
+	// far _untouched_. We can check with the prefetcher, if it can give us
+	// a trie which has the same root, but also has some content loaded into it.
+	// If so, use that one instead.
+	if s.prefetcher != nil {
+		s.prefetcher.Pause()
+		// We only want to do this _once_, if someone calls IntermediateRoot again,
+		// we shouldn't fetch the trie again
+		if s.originalRoot != (common.Hash{}) {
```

Comment:
Perhaps we can do the similar thing like storage trie, whenever we retrieve the resolved trie from the prefetcher, mark it as nil.

Reply:
You mean to just release the reference for GC, or do you mean for correctness?

Reply:
The cached tries should only be accessed for one time and then throw it away. E.g. storage trie.
So I mean for state trie it's same. If we fetch the state trie from `prefetcher` then mark it as nil.
(although the current code does the same logic)